### ISSUE TITLE
No results with WHERE IS NOT NULL on a datetime field

### ISSUE THREAD

**Issue by dphil** [dolthub/dolt#4199]
`WHERE field_name IS NULL` works on a datetime field, but `IS NOT NULL` does not. The latter returns no results, even if there are rows with non-null datetime values.

**Comment by dphil** [comment]
The issue appears to be if the column has some values that are NULL, and some that are not.
If a datetime column has no NULL values in it, a `WHERE field_name IS NOT NULL` query correctly returns all results. But apparently if there are any NULL results, than IS NOT NULL always returns an empty set.

**Comment by timsehn** [comment]
Nice find and repro. We'll crank out the fix.

**Comment by fulghum** [comment]
Thanks for reporting this one @dphil. I haven't been able to repro yet... do you see any other clues in your setup? 

Here's what I tried (with dolt version 0.40.27, as well as tip of main):
```
❯ dolt sql 
# Welcome to the DoltSQL shell.
# Statements must be terminated with ';'.
# "exit" or "quit" (or Ctrl-D) to exit.
> create table t (pk int primary key auto_increment, d datetime);
> insert into t (d) values (NOW()), (NULL), (NOW());
Query OK, 3 rows affected

> select * from t;
+----+----------------------------+
| pk | d                          |
+----+----------------------------+
| 1  | 2022-08-25 20:06:07.861593 |
| 2  | NULL                       |
| 3  | 2022-08-25 20:06:07.861593 |
+----+----------------------------+

> select * from t where d is NULL;
+----+------+
| pk | d    |
+----+------+
| 2  | NULL |
+----+------+

> select * from t where d is not NULL;
+----+----------------------------+
| pk | d                          |
+----+----------------------------+
| 1  | 2022-08-25 20:06:07.861593 |
| 3  | 2022-08-25 20:06:07.861593 |
+----+----------------------------+
```

If you can, let me know what Dolt version you saw this with and if possible, the output from `show create table` for that table might be helpful, too. I'll try a few more things to trigger this (like an index on the datetime field), but let me know if you find any other clues.

**Comment by fulghum** [comment]
oops! I spoke too soon! as soon as I tested with adding an index on the datetime field, I was able to repro. No more information needed. I'll debug through and see what's going on.

**Comment by fulghum** [comment]
I made some progress on this today, but I haven't found the root cause yet. I did find that for only the date/datetime/time types, the non-null values are being sorted *after* the null values in the index. This causes the index iterator to see a null value, then quit, because it thinks those nulls mean the end of data and it know it's looking for non-null values. I've just started debugging through how the index is created and looks like I need to keep digging a little deeper. I'll get back to this one tomorrow morning and keep you updated. 

Thanks again for reporting this one! We'll get it sorted out.

**Comment by fulghum** [comment]
I found the code that was causing the index ordering of null and non-null values to be inconsistent for datetime types and I've got a patch in place that is now making my repro test pass. I need to clean the patch up a little more, do a little more testing, and will have a PR out for review this afternoon.

**Comment by fulghum** [comment]
Got this one reviewed and merged into main. We'll get this fix out in the next release, early next week, and I'll follow up here one more time when it's out. 

Thanks for opening this issue so we could track the fix! 🙏

**Comment by fulghum** [comment]
Closing the loop here – this is now released in Dolt 0.40.30.
